pkg/controllers/work/apply_controller.go

```diff
@@ -383,13 +389,46 @@ func (r *ApplyWorkReconciler) applyUnstructured(ctx context.Context, gvr schema.
 	}
 
 	// We only try to update the object if its spec hash value has changed.
-	if manifestObj.GetAnnotations()[manifestHashAnnotation] != curObj.GetAnnotations()[manifestHashAnnotation] {
+	if manifestObj.GetAnnotations()[ManifestHashAnnotation] != curObj.GetAnnotations()[ManifestHashAnnotation] {
+		// we need to merge the owner reference between the current and the manifest since we support one manifest
+		// belong to multiple work so it contains the union of all the appliedWork.
+		manifestObj.SetOwnerReferences(mergeOwnerReference(curObj.GetOwnerReferences(), manifestObj.GetOwnerReferences()))
+		// record the raw manifest with the hash annotation in the manifest.
+		if err := setModifiedConfigurationAnnotation(manifestObj); err != nil {
+			return nil, ManifestNoChangeAction, err
+		}
+		annotations := manifestObj.GetAnnotations()
+		if annotations[LastAppliedConfigAnnotation] == "" {
+			klog.V(2).InfoS("using server side apply for manifest", "gvr", gvr, "manifest", manifestRef)
+			return r.applyObject(ctx, gvr, manifestObj)
+		}
```

Comment:
Why is this needed if the setModifiedConfigurationAnnotation returns if the annotation is set or not. It seems that the information is ready but we somehow checked it again.

Reply:
In the current implementation, setModifiedConfigurationAnnotation sets the lastAppliedConfig Annotation if the annotation size is valid if not it sets the  lastAppliedConfig annotation to "". The check is to ensure if we require a server side apply or three way merge patch